### PR TITLE
Refactor: move version, os, features from ClientYamlTestClient to ClientYamlTestExecutionContext

### PR DESCRIPTION
**Pull request by ldematte**
Just moving stuff around, no change in behaviour. 

Moving these fields showed how we are not handling them correctly in derived classes where multiple clusters are tested (ex: CCR), _but_ we want to separate refactor and change in behaviour.

The fix for derived classes will be in a follow-up PR.

### CODE REVIEW

**Comment by ldematte** [comment]
Failures in BwC tests are a known issue (https://github.com/elastic/elasticsearch/issues/103358) which is still unresolved unfortunately

File: qa/multi-cluster-search/src/test/java/org/elasticsearch/search/MultiClusterSearchYamlTestSuiteIT.java [hunk 1]
```diff
@@ -33,27 +36,36 @@ public static void determineRemoteClusterMinimumVersion() {
...
+         * Since the esVersion is used to skip tests in ESClientYamlSuiteTestCase, we also take into account the
+         * remote cluster version here and return it if it is lower than the local client version. This is used to
+         * skip tests if some feature isn't available on the remote cluster yet.
+         */
+        final Version commonEsVersion;
+        if (remoteEsVersion == null) {
+            commonEsVersion = esVersion;
+        } else {
+            commonEsVersion = remoteEsVersion.before(esVersion) ? remoteEsVersion : esVersion;
+        }
```
Comment (mosche) [inline qa/multi-cluster-search/src/test/java/org/elasticsearch/search/MultiClusterSearchYamlTestSuiteIT.java:57]: How about merging the null check into the 2nd statement?
```suggestion
        final Version commonEsVersion = remoteEsVersion != null && remoteEsVersion.before(esVersion) ? remoteEsVersion : esVersion;
```

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java [hunk 2]
```diff
@@ -179,30 +192,30 @@ public void initAndResetContext() throws Exception {
...
+        ClientYamlTestClient clientYamlTestClient,
         final Version esVersion,
         final Predicate<String> clusterFeaturesPredicate,
         final String os
     ) {
-        return new ClientYamlTestClient(
-            restSpec,
-            restClient,
-            hosts,
+        // TODO:
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java:200]: Remove the empty todo?

**Review by mosche — APPROVED** [review]
LGTM, just a nit regarding that empty todo

**Pull request merged**